### PR TITLE
added split and collapse operations

### PR DESCRIPTION
**Pull request by Zhouyuan-Chen**
Added split and collapse operations for tetrahedra mesh.

### CODE REVIEW

**Comment by Zhouyuan-Chen** [comment]
Just want to mention that seems Boundary Detect in 3D is not implemented. In this case, I didn't use the related invariant.

File: src/wmtk/operations/tet_mesh/TetSplit.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    : TetMeshOperation(m)
+    , TupleOperation(settings.invariants, t)
+    , m_settings{settings}
+{
+    assert(m_settings.are_invariants_initialized());
+}
+
+bool TetSplit::execute()
+{
+    Tuple first_split_ret = mesh().split_edge(input_tuple(), hash_accessor()).m_output_tuple;
```
Comment (daniel-zint) [inline src/wmtk/operations/tet_mesh/TetSplit.cpp:40]: You should not use atomic operations in here but `TetEdgeSplit` and `TetEdgeCollapse`.

Also, check the naming convention from the tri_mesh folder. `TetEdgeSplit` should only be called `EdgeSplit` and `TetEdgeCollapse` should be `EdgeCollapse`.

File: src/wmtk/operations/tet_mesh/TetSplit.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+bool OperationSettings<tet_mesh::TetSplit>::are_invariants_initialized() const
+{
+    return find_invariants_in_collection_by_type<ValidTupleInvariant>(invariants);
+}
+
+namespace tet_mesh {
+
+// TetSplit::TetSplit(Mesh& m, const Tuple& t, const OperationSettings<TetSplit>& settings)
```
Comment (daniel-zint) [inline src/wmtk/operations/tet_mesh/TetSplit.cpp:22]: Delete commented out code.

File: src/wmtk/operations/tet_mesh/TetSplit.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+Tuple TetSplit::return_tuple() const
+{
+    return m_output_tuple;
+}
+
+std::vector<Tuple> TetSplit::modified_primitives(PrimitiveType type) const
+{
+    if (type == PrimitiveType::Face) {
+        // TODO
```
Comment (daniel-zint) [inline src/wmtk/operations/tet_mesh/TetSplit.cpp:83]: if this is really a TODO, you want this code to throw an error if someone tries to run it.

**Review by daniel-zint — CHANGES_REQUESTED** [review]
looks almost good

File: src/wmtk/operations/tet_mesh/TetSplit.cpp (hunk 3 above)
Comment (Zhouyuan-Chen) [reply src/wmtk/operations/tet_mesh/TetSplit.cpp:83]: I copied this from tri_mesh. I am not sure what is this supposed to do. But definitely, we need this function...

Comment (daniel-zint) [reply src/wmtk/operations/tet_mesh/TetSplit.cpp:116]: Don't copy code without testing it and without knowing what it does. That can lead to very annoying debugging sessions later on. It is better if the code crashes or contains only half the functionality than having code that appears to the outside like it is looking the right thing.

**Comment by Zhouyuan-Chen** [comment]
okay, I implemented the primitive_modified function

**Comment by Zhouyuan-Chen** [comment]
I added the tests for the function primitive_modified.

**Review by daniel-zint — APPROVED** [review]
lgtm

**Pull request merged**